### ISSUE TITLE
Any document to compile the ins?

### ISSUE THREAD

**Issue by willard-yuan** [stylixboom/ins#1]
Hi stylixboom,

It's awesome that you implemented the paper "Total Recall: Automatic Query Expansion with a Generative Feature Model for Object Retrieval". I'm also very interested in object-based image retrieval and your project of the **ins**. But it's a little difficulty for me to compile it without any document guide. Would you give me some help to compile it?

**Comment by stylixboom** [comment]
Hi,

Thanks for your interest.
I implemented only the basic part of Total Recall, which is well known as an average query expansion (AQE). You can directly go to qe.h and qe.cpp under my ins project.

For the basic way to compile, I pushed the Makefile for ins and its related project. The dependency of the required libraries can be found within the Makefile. However, you need to modify lib/include paths for your environment.

I will make a short document on requirements and how to compile these libraries a bit later.

**Comment by willard-yuan** [comment]
Hi stylixboom,

It's very nice of you for you help. I'll try to compile it with the makefile later. I also do image retrieval research and it's about two years for me to dive into this area. Recently I try to write object-based image retrieval application with C plus plus( in fact, from the past two years I wrote so much code with Matlab ), and re-rank the  search result by RANSAC. I found some the RANSAC results are not so good. For example:

![image](https://cloud.githubusercontent.com/assets/5379711/9292624/2520121e-4437-11e5-95fa-6bfcbbb9f766.png)

![image](https://cloud.githubusercontent.com/assets/5379711/9292635/bdbe9054-4437-11e5-85ad-299c6bf67e37.png)

![image](https://cloud.githubusercontent.com/assets/5379711/9292641/1ccc48b6-4438-11e5-9e9a-5e52aa4beb88.png)

The top is good. However, the bottom is bad since most of good matches are removed. Have you faced the problem when do reranking, and Is it better for me to consider the 2-nn when reranking?

**Comment by stylixboom** [comment]
Hi willard-yuan,

In order to do re-ranking, we usually use the number of inlier point that verified in bi-directional.
What does it means?
Between the left and the right image, we run ransac to get the match points on the right image,
From the right one, then we do back-projection the matched points (by using homography matrix got from Ransac itself) to the left image, to check if they fall within the roi of the left image.
To answer your question, 2-nn might not necessary.

In additional,
How did you tune the Ransac parameter, that is what you may need to check.

**Comment by willard-yuan** [comment]
Hi stylixboom,

Thank you for spending time replying to me. I have got a general idea of the method verified in bi-directional. In my case, I don't want to use ROI, you know the ROI is often selected with a bounding box by manual or using saliency detection method, and it's not good from the view of Human-Computer Interaction. Using the saliency detection method will increase the running time and sometimes the area is not what we want.

In the re-ranking stage, I selected the key points as the matched points if they share the same visual words between the query and the candidates, Then I treat the number of inliers as the re-ranking score.

I find the code of [practical-object-instance-recognition](https://github.com/vedaldi/practical-object-instance-recognition) written by Andrea Vedaldi is very useful. The performance of [geometricVerification.m](https://github.com/vedaldi/practical-object-instance-recognition/blob/master/geometricVerification.m) is great, and I'm diving into rewriting it by C plus plus. I hope it's also helpful for you.

Thanks.
